specs/state-proof.md

```diff
@@ -0,0 +1,351 @@
+
+When state proof is generated, a random access record set is also produced with the global counter as each record's unique identifier, we called this set **bus mapping** because it acts like the bus in computer which transfers data between components. The bus mapping will be shared to EVM proof, then EVM proof can do random read-write access to anything inside the set at any time, with extra subset arguments to argue those accesses indeed form a subset of the one prodced by state proof.
+
+## Random Read-Write Data
+
+State proof maintains the read-write part of [random accessible data](./evm-proof.md#Random-Accessible-Data) of evm proof:
+
+| Target                                | Index             | Description                              |
+| ------------------------------------- | ----------------- | ---------------------------------------- |
+| [`AccountNonce`](#AccountNonce)       | `{address}`       | Account's nonce                          |
+| [`AccountBalance`](#AccountBalance)   | `{address}`       | Account's balance                        |
+| [`AccountCodeHash`](#AccountCodeHash) | `{address}`       | Account's code hash                      |
+| [`AccountStorage`](#AccountStorage)   | `{address}.{key}` | Account's storage as a key-value mapping |
+| [`CallState`](#CallState)             | `{id}.{enum}`     | Call's internal state                    |
+| [`CallStateStack`](#CallStateStack)   | `{id}.{index}`    | Call's stack as a encoded word array     |
+| [`CallStateMemory`](#CallStateMemory) | `{id}.{index}`    | Call's memory as a byte array            |
+
+The concatenation of **Target** and **Index** becomes the unique index for data. Each record will be attached with a global counter as their accessing timestamp, and the records are constraint to be in group by their unique index first and to be sorted by their global counter increasingly. Given the access to previous record, each target has their custom constraints, for example, `AccountNonce` always increase by 1, and values in `CallStateMemory` should fit in 8-bit.
+
+## Circuit Constraints
+
+The following part describes the custom constraints for each target by a python script. There are some common helper class and function:
+
+```python
+def range_lookup(value: int, range: str):
+    result = re.search('^([(\[])(.+),\s+(.+)([)\]]$)', range)
+    min = eval(result.group(2)) + (1 if result.group(1) == '(' else 0)
+    max = eval(result.group(3)) - (1 if result.group(4) == ')' else 0)
+    return min <= value and value <= max
```

Comment:
Are we defining the `range_lookup` function instead of using something like `assert min < value <= max` because we want to emphasize that we have to perform table lookup for this check?
An alternative is explicitly saying all `assert min < value <= max` check requires a table lookup then use that style everywhere.

Reply:
I was trying use the latter, but when value is allowed to be `0`, `min` becomes `-1`, which seems not intuitive. Maybe we could have convention `assert min <= value < max`, which seems to be more intuitive just like the range in most language.